File: reference-implementations/python/src/toml_schema/_definition.py
"""The Definition and Condition data model for a parsed schema node."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Pattern, Tuple

from ._types import SchemaType


@dataclass
class Condition:
    """An ``if`` selector: ``{ key = ..., equals = ... }`` or
    ``{ key = ..., in = [...] }``."""

    key: str
    has_equals: bool
    equals: Any = None
    in_values: Tuple[Any, ...] = ()


@dataclass
class Definition:
    """A single parsed schema definition node (a ``[types.*]`` or
    ``[elements.*]`` entry, or one of its nested child definitions).

    Public accessors mirror the other reference implementations:
    ``description``, ``deprecated``, ``default()``, and ``child(name)``.
    Internal validator state (references, constraints, sibling rules) is
    consumed by the schema/validator modules within this package.
    """

    name: str
    type_name: Optional[SchemaType] = None
    reference: str = ""
    description: str = ""
    item_reference: str = ""
    items: Tuple[str, ...] = ()
    optional: bool = False
    allowed_values: Tuple[Any, ...] = ()
    pattern: Optional[Pattern[str]] = None
    format: str = ""
    key_pattern: Optional[Pattern[str]] = None
    min: Any = None
    max: Any = None
    min_length: Optional[int] = None
    max_length: Optional[int] = None
    one_of: Tuple[str, ...] = ()
    any_of: Tuple[str, ...] = ()
    condition: Optional[Condition] = None
    then_reference: str = ""
    else_reference: str = ""
    all_of: Tuple[str, ...] = ()
    dependent_required: Dict[str, Tuple[str, ...]] = field(default_factory=dict)
    mutually_exclusive: Tuple[Tuple[str, ...], ...] = ()
    exactly_one: Tuple[Tuple[str, ...], ...] = ()
    unique_items: Optional[bool] = None
    default_value: Any = None
    has_default: bool = False
    deprecated: bool = False
    has_deprecated: bool = False
    children: Dict[str, "Definition"] = field(default_factory=dict)

    def default(self) -> Tuple[Any, bool]:
        """Returns the effective, non-materializing default annotation as
        ``(value, has_default)``."""
        return self.default_value, self.has_default

    def child(self, name: str) -> Optional["Definition"]:
        return self.children.get(name)

    def child_names(self) -> List[str]:
        return list(self.children.keys())

File: reference-implementations/typescript/src/definition.ts
import type { SchemaType } from "./builtins.js";
import type { StringFormat } from "./formats.js";
import type { TomlValue } from "./values.js";

/** A parsed `if` selector: `{ key = "...", equals = ... }` or `{ key = "...", in = [...] }`. */
export interface Condition {
  readonly key: string;
  readonly hasEquals: boolean;
  readonly equals?: TomlValue | undefined;
  readonly in: readonly TomlValue[];
}

/**
 * The internal, mutable-at-parse-time representation of a single schema
 * definition node (an element, a named type, or a nested child). This
 * mirrors the Go reference implementation's `Definition` struct field for
 * field; the public, read-only `Definition` class wraps a resolved instance
 * of this shape for consumers.
 */
export interface RawDefinition {
  name: string;
  typeName?: SchemaType | undefined;
  reference?: string | undefined;
  description?: string | undefined;
  itemReference?: string | undefined;
  items?: readonly string[] | undefined;
  optional: boolean;
  allowedValues?: readonly TomlValue[] | undefined;
  pattern?: RegExp | undefined;
  patternSource?: string | undefined;
  format?: StringFormat | undefined;
  keyPattern?: RegExp | undefined;
  keyPatternSource?: string | undefined;
  min?: TomlValue | undefined;
  max?: TomlValue | undefined;
  minLength?: number | undefined;
  maxLength?: number | undefined;
  oneOf?: readonly string[] | undefined;
  anyOf?: readonly string[] | undefined;
  condition?: Condition | undefined;
  thenReference?: string | undefined;
  elseReference?: string | undefined;
  allOf?: readonly string[] | undefined;
  dependentRequired?: Readonly<Record<string, readonly string[]>> | undefined;
  mutuallyExclusive?: readonly (readonly string[])[] | undefined;
  exactlyOne?: readonly (readonly string[])[] | undefined;
  uniqueItems?: boolean | undefined;
  defaultValue?: TomlValue | undefined;
  hasDefault: boolean;
  deprecated: boolean;
  children: Readonly<Record<string, RawDefinition>>;
}

/** Creates a string-keyed record without inherited Object prototype members. */
export function emptyRecord<T>(): Record<string, T> {
  return Object.create(null) as Record<string, T>;
}

/** Returns a shallow clone of `definition` with `overrides` applied. */
export function cloneDefinition(
  definition: RawDefinition,
  overrides: Partial<RawDefinition> = {},
): RawDefinition {
  return { ...definition, ...overrides };
}

/** Merges two dependentrequired maps (used when composing/resolving references). */
export function mergeDependentRequired(
  left: Readonly<Record<string, readonly string[]>> | undefined,
  right: Readonly<Record<string, readonly string[]>> | undefined,
): Record<string, readonly string[]> | undefined {
  if (!left && !right) return undefined;
  const merged = emptyRecord<string[]>();
  for (const [trigger, dependencies] of Object.entries(left ?? {})) {
    merged[trigger] = [...(merged[trigger] ?? []), ...dependencies];
  }
  for (const [trigger, dependencies] of Object.entries(right ?? {})) {
    merged[trigger] = [...(merged[trigger] ?? []), ...dependencies];
  }
  return merged;
}

/**
 * Public, read-only accessor over a resolved schema definition node.
 *
 * Instances are returned by {@link Schema.element}, {@link Schema.type}, and
 * {@link Definition.child}, always with inherited/effective annotations
 * (`description`, `deprecated`, `default`) already resolved through
 * references and `allof` composition.
 */
export class Definition {
  readonly #raw: RawDefinition;

  /** @internal */
  constructor(raw: RawDefinition) {
    this.#raw = raw;
  }

  /** @internal */
  get raw(): RawDefinition {
    return this.#raw;
  }

  /** The definition's simple name (its key within its parent table). */
  get name(): string {
    return this.#raw.name;
  }

  /** The effective `description`, inherited from a named type reference if unset locally. */
  get description(): string | undefined {
    return this.#raw.description;
  }

  /** The effective `deprecated` flag, propagated through references and `allof`. */
  get deprecated(): boolean {
    return this.#raw.deprecated;
  }

  /** Whether this definition declares an effective, non-materializing `default`. */
  hasDefault(): boolean {
    return this.#raw.hasDefault;
  }

  /** The effective `default` value, or `undefined` if none is declared. */
  default(): TomlValue | undefined {
    return this.#raw.defaultValue;
  }

  /** Whether the parent may omit this value. */
  get optional(): boolean {
    return this.#raw.optional;
  }

  /** Looks up a fixed child definition by name (for `table`/`collection` kinds). */
  child(name: string): Definition | undefined {
    const child = this.#raw.children[name];
    return child ? new Definition(child) : undefined;
  }

  /** Lists the names of all fixed child definitions. */
  childNames(): string[] {
    return Object.keys(this.#raw.children);
  }
}
